native/dtrevc3.go

```diff
@@ -0,0 +1,862 @@
+	b := work[n : n+n*ldb]
+
+	// Compute 1-norm of each column of strictly upper triangular part of T
+	// to control overflow in triangular solver.
+	norms[0] = 0
+	for j := 1; j < n; j++ {
+		norms[j] = 0
+		for i := 0; i < j; i++ {
+			norms[j] += math.Abs(t[i*ldt+j])
+		}
+	}
+
+	bi := blas64.Implementation()
+
+	var (
+		x [4]float64
+
+		iv int // Index of column in current block.
+		is int
+
+		// ip is used below to specify the real or complex eigenvalue:
+		//  ip == 0, real eigenvalue,
+		//        1, first  of conjugate complex pair (wr,wi),
+		//       -1, second of conjugate complex pair (wr,wi).
+		ip        int
+		iscomplex [nbmax]int // Stores ip for each column in current block.
+	)
+
+	if side == lapack.LeftEigVec {
+		goto leftev
+	}
+
+	// Compute right eigenvectors.
+
+	// For complex right vector, iv-1 is for real part and iv for complex
+	// part. Non-blocked version always uses iv=1, blocked version starts
+	// with iv=nb-1 and goes down to 0 or 1.
+	iv = max(2, nb) - 1
+	ip = 0
+	is = m - 1
+	for ki := n - 1; ki >= 0; ki-- {
+		if ip == -1 {
+			// Previous iteration (ki+1) was second of
+			// conjugate pair, so this ki is first of
+			// conjugate pair.
+			ip = 1
+			continue
+		}
+
+		if ki == 0 || t[ki*ldt+ki-1] == 0 {
+			// Last column or zero on sub-diagonal, so this
+			// ki must be real eigenvalue.
+			ip = 0
+		} else {
+			// Non-zero on sub-diagonal, so this ki is
+			// second of conjugate pair.
+			ip = -1
+		}
+
+		if howmny == lapack.SelectedEigVec {
+			if ip == 0 {
+				if !selected[ki] {
+					continue
+				}
+			} else if !selected[ki-1] {
+				continue
+			}
+		}
+
+		// Compute the ki-th eigenvalue (wr,wi).
+		wr := t[ki*ldt+ki]
+		wi := 0.0
+		if ip != 0 {
+			wi = math.Sqrt(math.Abs(t[ki*ldt+ki-1])) * math.Sqrt(math.Abs(t[(ki-1)*ldt+ki]))
+		}
+		smin := math.Max(ulp*(math.Abs(wr)+math.Abs(wi)), smlnum)
+
+		if ip == 0 {
+			// Real right eigenvector.
+
+			b[ki*ldb+iv] = 1
+			// Form right-hand side.
+			for k := 0; k < ki; k++ {
+				b[k*ldb+iv] = -t[k*ldt+ki]
+			}
+			// Solve upper quasi-triangular system:
+			//  [ T[0:ki,0:ki] - wr ]*X = scale*b.
+			for j := ki - 1; j >= 0; {
+				if j == 0 || t[j*ldt+j-1] == 0 {
+					// 1×1 diagonal block.
+					scale, xnorm, _ := impl.Dlaln2(false, 1, 1, smin, 1, t[j*ldt+j:], ldt,
+						1, 1, b[j*ldb+iv:], ldb, wr, 0, x[:], 2)
+					// Scale X[0,0] to avoid overflow when updating the
+					// right-hand side.
+					if xnorm > 1 && norms[j] > bignum/xnorm {
+						x[0] /= xnorm
+						scale /= xnorm
+					}
+					// Scale if necessary.
+					if scale != 1 {
+						bi.Dscal(ki+1, scale, b[iv:], ldb)
+					}
+					b[j*ldb+iv] = x[0]
+					// Update right-hand side.
+					bi.Daxpy(j, -x[0], t[j:], ldt, b[iv:], ldb)
+					j--
+				} else {
+					// 2×2 diagonal block.
+					scale, xnorm, _ := impl.Dlaln2(false, 2, 1, smin, 1, t[(j-1)*ldt+j-1:], ldt,
+						1, 1, b[(j-1)*ldb+iv:], ldb, wr, 0, x[:], 2)
+					// Scale X[0,0] and X[1,0] to avoid overflow
```

Comment:
I'm not sure I fully understand the question and what is unclear.

A possible answer: We multiply the corresponding column(s) of T by x[0] (x[1], ...) and subtract it from the right-hand side (in Daxpy below). `norms` holds the maximum norm of each column of T. So if the norm of x times the norm of the column of T is greater than bignum, then there is an element of T in that column that will overflow. Therefore in that case we normalize x to avoid the overflow. Scaling by `scale` comes from Dlaln2. Scaling of `scale` by `xnorm` is inevitable once we scale x[0] (x[1], ...) because we are scaling one component of the eigenvector so we must scale the previously computed components too and the right hand side (to scale the components not yet computed). And we are free to multiply the eigenvector by any non-zero number.